lib/ACL/ACLManager.php

```diff
@@ -54,40 +54,48 @@ private function getRootStorageId() {
 		return $this->rootStorageId;
 	}
 
-	private function pathsAreCached(array $paths): bool {
-		foreach ($paths as $path) {
-			if (!$this->ruleCache->hasKey($path)) {
-				return false;
-			}
-		}
-		return true;
-	}
-
 	/**
 	 * @param int $folderId
 	 * @param array $paths
 	 * @return (Rule[])[]
 	 */
 	private function getRules(array $paths): array {
-		if ($this->pathsAreCached($paths)) {
-			$rules = array_combine($paths, array_map(function (string $path) {
-				return $this->ruleCache->get($path);
-			}, $paths));
-		} else {
-			$rules = $this->ruleManager->getRulesForFilesByPath($this->user, $this->getRootStorageId(), $paths);
-			foreach ($rules as $path => $rulesForPath) {
+		// beware: adding new rules to the cache besides the cap
+		// might discard former cached entries, so we can't assume they'll stay
+		// cached, so we read everything out initially to be able to return it
+		$rules = array_combine($paths, array_map(function (string $path) {
+			return $this->ruleCache->get($path);
+		}, $paths));
+
+		$nonCachedPaths = [];
+		foreach ($paths as $path) {
+			if (!$this->ruleCache->hasKey($path)) {
```

Comment:
This seems redundant with the array_combine above, should be `$rules[$path] !== null` or `isset($rules[$path])` no?
`$nonCachedPaths` may even be computed by array_filter but not sure about the performance impact.